### PR TITLE
Issue 24 api key

### PR DESCRIPTION
**Pull request by mathewz12**
Creates an api key on R server startup that sends it to the perl server. Both sides stores the api key. Api key is required to be sent in every call from the R server to the perl server or else the call will be dropped.

### CODE REVIEW

File: inst/perl/lib/BiodiverseR.pm [hunk 1]
```diff
@@ -268,6 +325,27 @@ $log->debug("Called startup");
...
+        # $log->debug($perl_stored_api_key);
+        # $log->debug("Sent api key is:");
+        # $log->debug($sent_api_key);
+        
+        # Keys are undefined in Perl tests but I believe thats because its not calling anything from the R server so it makes sense there are no keys
+        $sent_api_key //= "";
+
+        $perl_stored_api_key //= "";
+
+        return "Stored api_key does not match api_key passed in"
```
Comment (shawnlaffan) [inline inst/perl/lib/BiodiverseR.pm:343]: Perhaps this should raise an exception on failure.  Then any callers would need to wrap the calls in ```eval{}``` and check for errors.  (We could also use [Feature::Compat::Try](https://metacpan.org/pod/Feature::Compat::Try) but that's for later).

**Comment by shawnlaffan** [comment]
Nearly all the perl tests should be failing since they do not set or pass the api_key.  

So it would seem the api_key checks are not yet kicking in?

**Comment by shawnlaffan** [comment]
Although perhaps it is better to keep the api key optional, in which case we can add some specific api-based perl tests.

File: inst/perl/lib/BiodiverseR.pm [hunk 2]
```diff
@@ -268,6 +325,27 @@ $log->debug("Called startup");
...
+        my $body_params = $c->req->json;
+        my $sent_api_key = $body_params->{api_key};
+        my $perl_stored_api_key = $api_key;
+        # $log->debug("Stored api key is:");
+        # $log->debug($perl_stored_api_key);
+        # $log->debug("Sent api key is:");
+        # $log->debug($sent_api_key);
+        
+        # Keys are undefined in Perl tests but I believe thats because its not calling anything from the R server so it makes sense there are no keys
+        $sent_api_key //= "";
```
[the remark this answers is not in the document]
Comment (mathewz12) [reply inst/perl/lib/BiodiverseR.pm]: Done as suggested.

File: inst/perl/lib/BiodiverseR.pm [hunk 3]
```diff
@@ -250,6 +299,14 @@ $log->debug("Called startup");
         return $c->render(json => {error => $e, result => defined $result});
     });
 
+    # Store the api_key
+    $r->post ('/api_key' => sub ($c) {
```
[the remark this answers is not in the document]
Comment (mathewz12) [reply inst/perl/lib/BiodiverseR.pm:342]: Added this check right before we set the api key.

File: inst/perl/lib/BiodiverseR.pm (hunk 1 above)
Comment (mathewz12) [reply inst/perl/lib/BiodiverseR.pm]: I believe I have done what you asked. Every single call to check the api key is done through eval now.

**Comment by mathewz12** [comment]
> Nearly all the perl tests should be failing since they do not set or pass the api_key.
> 
> So it would seem the api_key checks are not yet kicking in?

Initially the perl tests were all failing but I made them all pass because whenever the api key was undefined it gets changed to an empty string. However, now that I have added the changes you have suggested the perl tests should go back to failing. The R tests should all pass though, including the one I created for the api key.

**Comment by mathewz12** [comment]
> Although perhaps it is better to keep the api key optional, in which case we can add some specific api-based perl tests.

I am not quite too sure how to make the api key optional. Don't we have to either send in the api key and the server checks for it or we don't use an api key at all?

**Comment by shawnlaffan** [comment]
This is looking good.  

I had a look at updating the perl tests and it made me think that the caller setting the api key is not the best design.  

It might be better to have the server generate the api key and then allow the caller one chance to get it from the server via the api_key path.  

The Digest::SHA module can be used for this (or something else).  https://metacpan.org/pod/Digest::SHA

Ideally we would also send the api key via the headers to avoid parsing large json documents only to drop them, but that can be handled under another issue.

**Comment by mathewz12** [comment]
> This is looking good.
> 
> I had a look at updating the perl tests and it made me think that the caller setting the api key is not the best design.
> 
> It might be better to have the server generate the api key and then allow the caller one chance to get it from the server via the api_key path.
> 
> The Digest::SHA module can be used for this (or something else). https://metacpan.org/pod/Digest::SHA
> 
> Ideally we would also send the api key via the headers to avoid parsing large json documents only to drop them, but that can be handled under another issue.

Updated to have the Perl server generate the key instead of R. Then the Perl server will send the API key once to the R server. Afterwards all calls made by the R server to the Perl server will have the API key attached.

**Pull request merged**